box/src/main/scala/blobstore/box/BoxStore.scala

```diff
@@ -0,0 +1,232 @@
+
+    val release: ((OutputStream, InputStream)) => F[Unit] = ios => F.delay {
+      ios._1.close()
+      ios._2.close()
+    }
+
+    Stream.bracket(init)(consume, release)
+  }
+
+  /**
+    * Creates a BoxFolder at this path and all folders along this path.
+    * If the path already exists, this will simply traverse the path and return the folder at this path.
+    * NOTE: this method makes Box API calls
+    * @param parentFolder
+    * @param pathParts
+    * @return a BoxFolder at this path
+    */
+  def putFolderAtPath(parentFolder: BoxFolder, pathParts: List[String]): BoxFolder = {
+    pathParts match {
+      case Nil => parentFolder
+      case head :: tail =>
+        val matchingItem = parentFolder.getChildren.asScala
+          .find(_.getName.equalsIgnoreCase(head))
+          .getOrElse(parentFolder.createFolder(head))
+
+        if (!matchingItem.getResource.isInstanceOf[BoxFolder]) {
+          throw new Exception("Item exists at path but was not folder")
+        } else {
+          val folder = matchingItem.getResource.asInstanceOf[BoxFolder]
+          putFolderAtPath(folder, tail)
+        }
+    }
+  }
+
+  def splitPath(path: Path): (List[String], String) = {
+    val fullPath = path.root :: path.key.split("/").toList
+    val pathToParentFolder = fullPath.dropRight(1)
+    val key = fullPath.takeRight(1).head
+    (pathToParentFolder, key)
+  }
+
+  /**
+    * Provides a Sink that writes bytes into the provided path. See [[StoreOps.PutOps]] for convenient put String
+    * and put file methods.
+    *
+    * It is highly recommended to provide [[Path.size]] when writing as it allows for optimizations in some store.
+    *
+    * @param path to put
+    * @return sink of bytes
+    */
+  // TODO: check for existence of file and delete it?
+  override def put(path: Path): Sink[F, Byte] = { in =>
+    val pathSplit = splitPath(path)
+
+    val init: F[(OutputStream, InputStream, BoxFolder)] = F.delay {
+      val parentFolder = putFolderAtPath(rootFolder, pathSplit._1)
+      val os = new PipedOutputStream()
+      val is = new PipedInputStream(os)
+      (os, is, parentFolder)
+    }
+
+    val consume: ((OutputStream, InputStream, BoxFolder)) => Stream[F, Unit] = ios => {
+      val putToBox = Stream.eval(F.delay({
+        ios._3.uploadFile(ios._2, pathSplit._2)
+        ()
+      }))
+      val writeBytes = _writeAllToOutputStream1(in, ios._1).stream ++ Stream.eval(F.delay(ios._1.close()))
+
+      putToBox concurrently writeBytes
+    }
+
+    val release: ((OutputStream, InputStream, BoxFolder)) => F[Unit] = ios => F.delay {
+      ios._2.close()
+      ios._1.close()
+    }
+
+    Stream.bracket(init)(consume, release)
+  }
+
+  /**
+    * Moves bytes from srcPath to dstPath. Stores should optimize to use native move functions to avoid data transfer.
+    *
+    * @param src path
+    * @param dst path
+    * @return F[Unit]
+    */
+  override def move(src: Path, dst: Path): F[Unit] = {
+    F.delay(
```

Comment:
This F.delay can be moved to the function signature -- same goes for the other ones
e.g.: ``F[Unit] = F.delay {..``